src/boot/efi/stub.c

```diff
@@ -253,29 +254,113 @@ static EFI_STATUS load_addons_from_dir(
 
 }
 
-static EFI_STATUS cmdline_append_and_measure_addons(
+static void cmdline_append_and_measure_addons(
+                char16_t *cmdline,
+                char16_t **cmdline_append,
+                bool *ret_parameters_measured) {
+
+        _cleanup_free_ char16_t *tmp = NULL;
+        bool m = false;
+
+        assert(ret_parameters_measured);
+        assert(cmdline_append);
+
+        mangle_stub_cmdline(cmdline);
+
+        if (isempty(cmdline))
+                return;
+
+        (void) tpm_log_load_options(cmdline, &m);
+        *ret_parameters_measured = m;
+
+        tmp = TAKE_PTR(*cmdline_append);
+        *cmdline_append = xasprintf("%ls%ls%ls", strempty(tmp), isempty(tmp) ? u"" : u" ", cmdline);
+}
+
+static void dtb_install_addons(
+                struct devicetree_state *dt_state,
+                void **dt_bases,
+                size_t *dt_sizes,
+                size_t n_dts,
+                bool *ret_parameters_measured) {
+
+        int parameters_measured = -1;
+        EFI_STATUS err;
+
+        assert(dt_state);
+        assert(n_dts == 0 || (dt_bases && dt_sizes));
+        assert(ret_parameters_measured);
+
+        for (size_t i = 0; i < n_dts; ++i) {
+                err = devicetree_install_from_memory(dt_state, dt_bases[i], dt_sizes[i]);
+                if (err != EFI_SUCCESS)
+                        log_error_status(err, "Error loading global addon devicetree, ignoring: %m");
+                else {
+                        _cleanup_free_ char16_t* description = NULL;
+                        bool m = false;
+
+                        description = xasprintf("DTB addon #%zu", i);
+
+                        err = tpm_log_event(
+                                        TPM2_PCR_KERNEL_CONFIG,
+                                        POINTER_TO_PHYSICAL_ADDRESS(dt_bases[i]),
+                                        dt_sizes[i],
+                                        description,
+                                        &m);
```

Comment:
so the way we so far did TPM event logs is terrible, because it makes is very hard to recognize stuff in the event logs, and that's important for systemd-pcrlock and similar tools.

Given you are measuring something we haven't measured before, let's please fix this, and measure things in a more useful, recognizable way.

Specifically this means tpm_log_event() should be improved to optionally not measure things as an EV_IPL event, but a "tagged event", i.e. EV_EVENT_TAG, with a recognizable tag. See how the kernel does this via INITRD_EVENT_TAG_ID + LOAD_OPTIONS_EVENT_TAG_ID.

And then, in then in the event data say the file name of what you are measuring here.

Reply:
instead of improving tpm_log_event() maybe add an alternative call tpm_log_event_tagged() or so

Reply:
(and add a comment in tpm_log_event() it should not be used for new measurements)

Reply:
I'll use the filename for the description. While I agree that we should fix how we do event logs, testing this PR is already difficult and tedious as it is - only partially works on arm64 which is slow as hell to build plus uboot which causes a kernel panic, so the only way to have some bare idea whether things are somewhat working is with debug prints. So I am not going to refactor the event logs stuff here, as it's a generic problem that affects the rest of the stub/loader too, and we can do it separately and on a more sane and usable environment

Reply:
sure, this can happen in preparatory step. But note that there's plenty of code that uses our existing measurements already, hence for all the stuff we already measure we cannot really change the records we generate. Which is very unfortunate, but people have learnt to live with it.

But for the future we should do better, hence all new measurements shoudl use EV_EVENT_TAG event log records

Reply:
> only partially works on arm64 which is slow as hell to build

Are you building in a VM/qemu-user chroot? I've recently learned how to do debian multiarch. Then you can easily cross-build with native speed.

> so the only way to have some bare idea whether things are somewhat working is with debug prints

Debugging sd-boot is possible: tools/debug-sd-boot.sh. Also, you may wanna build with `-Db_sanitize=undefined` to make sure there isn't any UB involved on our side.